test/Plutip/UtxoDistribution.purs

```diff
@@ -0,0 +1,214 @@
+  )
+import Contract.Transaction
+  ( TransactionInput
+  , TransactionOutput(TransactionOutput)
+  )
+import Contract.Utxos (utxosAt)
+import Contract.Value (Value, lovelaceValueOf)
+import Contract.Wallet (KeyWallet, withKeyWallet)
+import Control.Lazy (fix)
+import Data.Array (foldl, zip)
+import Data.BigInt (BigInt)
+import Data.BigInt as BigInt
+import Data.FoldableWithIndex (foldlWithIndex)
+import Data.List as List
+import Data.Map as Map
+import Data.Maybe (isJust)
+import Data.Newtype (unwrap, wrap)
+import Data.NonEmpty ((:|))
+import Data.Traversable (for_)
+import Data.Tuple.Nested (type (/\), (/\))
+import Effect.Class (liftEffect)
+import Effect.Exception (throw)
+import Mote (group, test)
+import Plutip.Types (InitialUTxOsWithStakeKey(InitialUTxOsWithStakeKey))
+import Plutip.UtxoDistribution (encodeDistribution, keyWallets)
+import Plutus.Types.Transaction (Utxo)
+import Test.Plutip.Common (config, privateStakeKey)
+import Test.QuickCheck (class Arbitrary, arbitrary)
+import Test.QuickCheck.Gen (Gen, arrayOf, chooseInt, frequency, randomSample')
+import TestM (TestPlanM)
+import Type.Prelude (Proxy(Proxy))
+
+suite :: TestPlanM Unit
+suite = group "Plutip UtxoDistribution" do
+  distrs <- liftEffect $ randomSample' 5 arbitrary
+  for_ distrs $ \distr ->
+    test
+      ( "runPlutipContract: stake key transfers with random distribution: "
+          <> show distr
+      )
+      $
+        withArbUtxoDistr
+          distr
+          \randDistr -> runPlutipContract config randDistr $
+            checkUtxoDistribution randDistr
+
+checkUtxoDistribution
+  :: forall distr wallet (r :: Row Type)
+   . UtxoDistribution distr wallet
+  => distr
+  -> wallet
+  -> Contract r Unit
+checkUtxoDistribution distr wallets = do
+  let
+    walletsArray = keyWallets (Proxy :: Proxy distr) wallets
+    walletUtxos = encodeDistribution distr
+  for_ walletsArray assertUtxosAtPlutipWalletAddress
+  assertCorrectDistribution $ zip walletsArray walletUtxos
+
+-- TODO: minimum value of 1 ada is hardcoded, tests become flaky below
+-- that value. Ideally this shouldn't be hardcoded
+genInitialUtxo :: Gen InitialUTxOs
+genInitialUtxo = map (BigInt.fromInt >>> (_ * BigInt.fromInt 1_000_000))
+  <$> arrayOf (chooseInt 1 1000)
+
+instance Arbitrary ArbitraryUtxoDistr where
+  arbitrary = fix \_ -> frequency <<< wrap $
+    (1.0 /\ pure UDUnit) :|
+      List.fromFoldable
+        [ 2.0 /\ (UDInitialUtxos <$> genInitialUtxo)
+        , 2.0 /\
+            ( UDInitialUtxosWithStake <$>
+                ( InitialUTxOsWithStakeKey
+                    <$> (pure privateStakeKey)
+                    <*> genInitialUtxo
+                )
+            )
+        , 4.0 /\ (UDTuple <$> arbitrary <*> arbitrary)
+        ]
+
+data ArbitraryUtxoDistr
+  = UDUnit
+  | UDInitialUtxos InitialUTxOs
+  | UDInitialUtxosWithStake InitialUTxOsWithStakeKey
+  | UDTuple ArbitraryUtxoDistr ArbitraryUtxoDistr
+
+instance Show ArbitraryUtxoDistr where
+  show = case _ of
```

Comment:
We want `Show` instances to be as close to code as possible. Please include literal constructor names in the strings. If you want to pretty-print the type, use a standalone function.